1. Реализация стека

VitaliyIliyushenkov opened the pull request:
new file:   stack_app/stack_app/App.config
	new file:   stack_app/stack_app/Form1.Designer.cs
	new file:   stack_app/stack_app/Form1.cs
	new file:   stack_app/stack_app/Form1.resx
	new file:   stack_app/stack_app/MyStack.cs
	new file:   stack_app/stack_app/Program.cs
	new file:   stack_app/stack_app/Properties/AssemblyInfo.cs
	new file:   stack_app/stack_app/Properties/Resources.Designer.cs
	new file:   stack_app/stack_app/Properties/Resources.resx
	new file:   stack_app/stack_app/Properties/Settings.Designer.cs
	new file:   stack_app/stack_app/Properties/Settings.settings
	new file:   stack_app/stack_app/stack_app.csproj

VitaliyIliyushenkov committed:
new file:   stack_app/stack_app.sln
	new file:   stack_app/stack_app/App.config
	new file:   stack_app/stack_app/Form1.Designer.cs
	new file:   stack_app/stack_app/Form1.cs
	new file:   stack_app/stack_app/Form1.resx
	new file:   stack_app/stack_app/MyStack.cs
	new file:   stack_app/stack_app/Program.cs
	new file:   stack_app/stack_app/Properties/AssemblyInfo.cs
	new file:   stack_app/stack_app/Properties/Resources.Designer.cs
	new file:   stack_app/stack_app/Properties/Resources.resx
	new file:   stack_app/stack_app/Properties/Settings.Designer.cs
	new file:   stack_app/stack_app/Properties/Settings.settings
	new file:   stack_app/stack_app/stack_app.csproj

VasilevsWork commented on stack_app/stack_app/Form1.cs:
```diff
@@ -0,0 +1,53 @@
+            listBox1.Items.Clear();
+            for (int i = mystack.Count()-1; i >= 0; i--) 
+            {
+                listBox1.Items.Add(mystack.Value(i));
```
Исключительная ситуация не обработана.

VitaliyIliyushenkov replied:
Не понял о какой ситуации идет речь, переполнение стека и ситуации с извлечением эл-та из пустого стека, обработаны.

        public void Push(T item)
        {
            if (_currentIndex == Capacity)
            {
                throw new Exception("Stack is full");
            }

            _dataArray[_currentIndex] = item;
            _currentIndex++;
        }
-------------------------------------------------------------------------------------------------------------------------------------

public T Peek()
{
    // если данных нет, выбрасываем исключение
    if (Count == 0)
    {
        throw new Exception("Stack is empty");
    }

    return _dataArray[_currentIndex - 1];
}

// извлечение элемента
public T Pop()
{
    if (Count == 0)
    {
        throw new Exception("Stack is empty");
    }
    var item = Peek();
    _currentIndex--;

    return item;
}

VasilevsWork replied:
Исключения считается обработанным, если оно было помещено и поймано в блоке `try/catch`

VitaliyIliyushenkov replied:
Я добавил блоки try/catch в Form1 как и было указано:

 private void numericUpDown1_ValueChanged(object sender, EventArgs e)
 {
     try { mystack = new MyStack<int>((int)numericUpDown1.Value);}
     catch  {throw new Exception("Filling error"); }
 }
 private void push_btn_Click(object sender, EventArgs e)
 {
     try { mystack.Push(Convert.ToInt32(textBox1.Text)); }
     catch {throw new Exception("Transmission error"); }
     listBox1_Refresh();
 }

VasilevsWork commented on stack_app/stack_app/Form1.cs:
```diff
@@ -0,0 +1,53 @@
+        //Задание стека
+        private void Form1_Load(object sender, EventArgs e)
+        {
+            mystack = new MyStack<int>(8);
```
Размер стека задается пользователем.

VitaliyIliyushenkov replied:
Видимо прослушал. В задании указано, что размер задается в конструкторе 
![image](https://github.com/VitaliyIliyushenkov/SAOD_2255_Iliushenkov_Vitaliy/assets/125152691/9202eac6-a1ba-4524-a42e-9d6c58b3fbfd)

Добавил NumericUpDown, при изменении значения которого создается стек с указанным в нем значением.

        //Задание стека
        private void numericUpDown1_ValueChanged(object sender, EventArgs e)
        {
            mystack = new MyStack<int>((int)numericUpDown1.Value);
        }

VasilevsWork replied:
Отлично.

VasilevsWork commented on stack_app/stack_app/MyStack.cs:
```diff
@@ -0,0 +1,91 @@
+    {
+        private int _currentIndex; // текущий размер стека
+        private T[] _dataArray; // данные стека
+        private int _stackMaxSize; //максимальное количество элементов в стеке
```
Можно реализовать без использования переменной `_stackMaxSize`.

VitaliyIliyushenkov replied:
Убрал данную переменную заменив на свойство Capacity:

public int Capacity
{
    get { return _dataArray.Length; }
    set { _dataArray = new T[value]; }
}

VasilevsWork commented on stack_app/stack_app/MyStack.cs:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        // метод для получения размера стека
+        public int Count()
```
Желательно реализовать, `Count` и `Capacity` как свойства.

VitaliyIliyushenkov replied:
Пожелания услышал, реализовал эти методы как свойства:

public int Count
{
    get { return _currentIndex; }
    set { _currentIndex = value; }
}

public int Capacity
{
    get { return _dataArray.Length; }
    set { _dataArray = new T[value]; }
}

VasilevsWork commented on stack_app/stack_app/MyStack.cs:
```diff
@@ -0,0 +1,91 @@
+        {
+            // если текущий размер равен максимальному, то смещаем стек на одну позицию
+            // при этом первый элемент удаляется
+            if (_currentIndex == _stackMaxSize)
+            {
+                RebuildData();
+            }
```
В случае превышения допустимого количества элементов следует выбрасывать исключение.

VitaliyIliyushenkov replied:
Исправлено на:

        // добавление нового элемента
        public void Push(T item)
        {
            if (_currentIndex == Capacity)
            {
                throw new Exception("Stack is full");
            }

            _dataArray[_currentIndex] = item;
            _currentIndex++;
        }

VasilevsWork commented on stack_app/stack_app/MyStack.cs:
```diff
@@ -0,0 +1,91 @@
+            _currentIndex = _stackMaxSize - 1;
+        }
+
+        public T Value(int i)
+        {
+            return _dataArray[i];
+        }
```
Данный метод не требуется, т.к. в стеке не предполагается наличие индексов для внешнего пользователя.

VasilevsWork replied:
Нужно было реализовать метод `Values()`, возвращающий актуальные элементы стека.

VitaliyIliyushenkov replied:
Реализован метод Values() возвращающий массив эл-ов нашего стека:

        public T[] Values()
        {
            return _dataArray;
        }

VasilevsWork replied:
Метод должен возвращать массив только с актуальными элементами.

VasilevsWork commented on stack_app/stack_app/MyStack.cs:
```diff
@@ -0,0 +1,91 @@
+            _currentIndex = 0;
+        }
+
+        private void RebuildData()
+        {
+            var newData = new T[_stackMaxSize];
+            for (var i = 1; i < _dataArray.Length; i++)
+            {
+                newData[i - 1] = _dataArray[i];
+            }
+
+            _dataArray = newData;
+            _currentIndex = _stackMaxSize - 1;
+        }
```
Данные метод не должен требоваться для реализации

VitaliyIliyushenkov replied:
Данный метод был убран в новой реализации.

VasilevsWork reviewed (commented):
Есть ряд недочетов реализации. Еще раз ознакомьтесь с заданием.

VitaliyIliyushenkov committed:
Доработки

VasilevsWork commented on stack_app/stack_app/rbacSsHrnwk.jpg:
Не до конца понимаю, как это используется в рамках работы приложения?

VasilevsWork commented on stack_app/stack_app/MyStack.cs:
```diff
@@ -0,0 +1,82 @@
+        public int Count
+        {
+            get { return _currentIndex; }
+            set { _currentIndex = value; }
```
`set` убрать или сделать `private`. Свойство только для чтения.

VitaliyIliyushenkov replied:
Сделал свойство private

private int Count
{
    get { return _currentIndex; }
    set { _currentIndex = value; }
}

VasilevsWork commented on stack_app/stack_app/MyStack.cs:
```diff
@@ -0,0 +1,82 @@
+        public int Capacity
+        {
+            get { return _dataArray.Length; }
+            set { _dataArray = new T[value]; }
```
`set` убрать или сделать `private`. Свойство только для чтения.

VitaliyIliyushenkov replied:
Сделал свойства private

private int Capacity
{
    get { return _dataArray.Length; }
    set { _dataArray = new T[value]; }
}

VasilevsWork replied:
Но получение значения не должно быть приватным. Можно сделать отдельно `set` приватным.

VasilevsWork reviewed (commented):
Все еще есть ряд вопросов по реализации

VitaliyIliyushenkov committed:
modified:   stack_app/stack_app/Form1.cs
	modified:   stack_app/stack_app/MyStack.cs
	deleted:    stack_app/stack_app/ugSZ-pZkJM4.jpg

VitaliyIliyushenkov commented:
Метод Values теперь возвращает массив с актуальными элементами без 0.

VitaliyIliyushenkov committed:
Delete stack_app/stack_app/rbacSsHrnwk.jpg

VasilevsWork commented on stack_app/stack_app/Form1.cs:
```diff
@@ -0,0 +1,69 @@
+
+        private void pop_btn_Click(object sender, EventArgs e)
+        {
+            textBox1.Text = mystack.Pop().ToString();
```
Нужно добавить `try/catch` везде, где могут быть исключения.

VitaliyIliyushenkov replied:
Исправленно на:
private void push_btn_Click(object sender, EventArgs e)
{
    try
    {
        mystack.Push(Convert.ToInt32(textBox1.Text));
    }
    catch (FormatException)
    {
        MessageBox.Show("Введите число");
    }
    catch (Exception)
    {
        MessageBox.Show("Произошло переполнение стека");
    }
    
    listBox1_Refresh();
}
и
private void pop_btn_Click(object sender, EventArgs e)
{
    try
    {
        textBox1.Text = mystack.Pop().ToString();
        listBox1_Refresh();
    }
    catch(Exception)
    {
        MessageBox.Show("Cтек пустой");
    }
}

private void peek_btn_Click(object sender, EventArgs e)
{
    try
    {
        textBox1.Text = mystack.Peek().ToString();
    }
    catch (Exception)
    {
        MessageBox.Show("Cтек пустой");
    }
}
Убрана проверка исключения при создании нового стека:
private void numericUpDown1_ValueChanged(object sender, EventArgs e)
{
        mystack = new MyStack<int>((int)numericUpDown1.Value);
        listBox1_Refresh();
}

VasilevsWork commented on stack_app/stack_app/Form1.cs:
```diff
@@ -0,0 +1,69 @@
+        //Задание стека
+        private void numericUpDown1_ValueChanged(object sender, EventArgs e)
+        {
+            try
+            {
+                mystack = new MyStack<int>((int)numericUpDown1.Value);
+            }
+            catch 
+            {
+                throw new Exception("Filling error");
+            }
```
Тут не должно быть исключений, по идее.

VitaliyIliyushenkov replied:
Убрана проверка исключения при создании нового стека:
private void numericUpDown1_ValueChanged(object sender, EventArgs e)
{
mystack = new MyStack((int)numericUpDown1.Value);
listBox1_Refresh();
}
Возможно исключение, что мы не задали размер массиву, исправил это путем задания numericUpDown минимального значения = 1;

VasilevsWork commented on stack_app/stack_app/Form1.cs:
```diff
@@ -0,0 +1,69 @@
+            }
+            catch
+            {
+                throw new Exception("Transmission error");
```
Нужно выводить пользователю сообщение, а не генерировать новое исключение.

VitaliyIliyushenkov replied:
Исправлено, добавлены строки вызывающие MessageBox

VasilevsWork reviewed (commented):
Требуются еще доработки

VitaliyIliyushenkov committed:
modified:   stack_app/stack_app/Form1.Designer.cs
	modified:   stack_app/stack_app/Form1.cs
	modified:   stack_app/stack_app/MyStack.cs

VasilevsWork reviewed (approved):
Задание принимается. Поправьте только все же get/set для Count и Capacity.

VitaliyIliyushenkov committed:
modified:   stack_app/stack_app/MyStack.cs
